modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/dataset_display/DatasetDisplay.js

```diff
@@ -95,7 +95,7 @@ function DatasetDisplay({
 		...currentViewProps
 	} = activeView;
 
-	const selectable = bulkActions?.length && selectedItemsKey;
+	const selectable = bulkActions?.length > 0 && selectedItemsKey;
```

Comment:
I gather this is just a stylistic change with no behavioral effect, right? (I can't imagine `length` would have been negative before.)

Reply:
It actually fixes some issues where zeros where being rendered on the page.

Reply:
I see, because:

- `0 && key` is `0` (React will render `0`).
- `false && key` is `false` (React will render nothing).

So if what you really want here is a boolean, would be better to cast:

```js
const selectable = !!(bulkActions?.length && selectedItemsKey);
```

When I look to see how `selectable` is being used I see:

- It is [passed into `ManagementBar` here](https://github.com/brunobasto/liferay-portal/blob/c288a3e5729ee993f3cabca9698b34af35ccd066/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/dataset_display/DatasetDisplay.js#L319), but [`ManagementBar` ignores that prop](https://github.com/brunobasto/liferay-portal/blob/c288a3e5729ee993f3cabca9698b34af35ccd066/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/dataset_display/management_bar/index.js#L23-L35) — we should probably fix this.
- It's also [passed as a context value](https://github.com/brunobasto/liferay-portal/blob/c288a3e5729ee993f3cabca9698b34af35ccd066/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/dataset_display/DatasetDisplay.js#L437).
- Looks like it is [expected to be boolean](https://github.com/wincent/liferay-portal/blob/master/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/dataset_display/DatasetDisplayContext.js#L28).
- I can see it getting read out of context [here](https://github.com/liferay/liferay-portal/blob/b16ea7116ee08e54f1afb1834b83db7cf54d8e1d/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/dataset_display/views/table/Table.js#L131).
- And from there it makes its way [here](https://github.com/liferay/liferay-portal/blob/b16ea7116ee08e54f1afb1834b83db7cf54d8e1d/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/dataset_display/views/table/TableHeadRow.js#L210), where I expect the `selectable && whatever` ends up rendering `0` when `selectable` is `0`.

Reply:
> So if what you really want here is a boolean, would be better to cast:
> 
> ```js
> const selectable = !!(bulkActions?.length && selectedItemsKey);
> ```

Hmm. Is that so that we avoid eventually doing `undefined > 0`? Otherwise I intuitively found it more natural this way. I like double negations for casting, but in this case it didn't occurred to me.

Sometimes I try to avoid it when I can get away without it because deep down I feel like there will be a tiny tiny performance overhead because of it. I looked it up today and it seems it's probably [as fast as it gets](https://www.sitepoint.com/javascript-double-negation-trick-trouble/) which thinking about it is no surprise because negation is such a low level operation (and the double negation can probably be optimised having it's own instruction or something).

Another reason I sometimes hold back is because it doesn't seem to be obvious for some people (I've had to explain why that's used quite a few times). I've also seem people use `Boolean(expression)`. What's your take on this? Either way is fine? Avoid Boolean() to avoid the overhead of a function call?

Reply:
I think in terms of JS "idioms", these two are common enough to warrant usage:

- `!!` to coerce to boolean in most cases.
- The exception when I'd use `Boolean` is for filtering; eg. `listOfStuff.filter(Boolean)`.

The `!!` idiom will be familiar to people coming to JS from other languages, and for people who are just starting out, it is easily learned.

Good point about the speed, and I'm glad you found a resource confirming that it the cost is insignificant. One of the reasons for that is that engine developers spend time optimizing "idiomatic" constructs, so it _pays to be idiomatic_ in that sense, not just for conformity to community style but also for performance.

When we move to TypeScript people are going to have to get more familiar with `!!` anyway, because in a situation like this one the compiler will know that `selectable` is supposed to have type `boolean`, and it will complain unless you assign a boolean:

- ❌ type `false | string` (ie. what `bulkActions?.length > 0 && selectedItemsKey` is, assuming the key is a string).
- ❌ type `number | string` (ie. what `bulkActions?.length && selectionItemsKey` is).
- ✅ type `boolean` (ie. what `!!(anything at all)` is).

Reply:
Fixed [here](https://github.com/liferay-frontend/liferay-portal/pull/257/commits/1871d6bf791db0734d55014b8f3b20185c336b5e) and [here](https://github.com/liferay-frontend/liferay-portal/pull/257/commits/aad347700e997dfc4ce6794c034fd54e85df76e1)